### DISCUSSION TITLE
auto retry

### DISCUSSION THREAD

**Discussion by francescodifilippo** [cline/cline#213] [Feature Requests]
In case of exceeding the rate limit it could be interesting to have an auto retry that waits the necessary time.
See the error:
429 {"type":"error","error":{"type":"rate_limit_error","message":"Number of request tokens has exceeded your per-minute rate limit (https://docs.anthropic.com/en/api/rate-limits); see the response headers for current usage. Please reduce the prompt length or the maximum tokens requested, or try again later. You may also contact sales at https://www.anthropic.com/contact-sales to discuss your options for a rate limit increase."}}

**Comment by CiberNin** [comment]
It should even be possible to implement an (actually accurate! ) loading bar.

**Comment by saoudrizwan** [comment]
For anybody dealing with 429 rate limit issues, I suggest using OpenRouter–they've just added prompt caching and have better rate limits.

**Reply by kam1092** [reply]
Auto retry with a timer  retry every 90 seconds to satisfy the 60 second rule of anthropic

**Comment by SinghChinmay** [comment]
Implement auto-retry logic: If a rate limit error occurs (429 with `rate_limit_error` message), retry the request after one minute. Add a new setting option, `autoRetry`, as a Boolean to enable or disable this feature.

**Comment by slavakurilyak** [comment]
Does `Auto-approve` or `Max Requests` help with `rate_limit_error`?

![CleanShot 2024-12-29 at 12 05 56](https://github.com/user-attachments/assets/8ebccf29-269e-4491-8408-fb1a237eef8f)

**Reply by alwinraju** [reply]
No it doesn't unfortunately. The max requests is how many requests Cline sends before your approval is required, not the speed at which it sends requests/tokens.

**Comment by perrosnk** [comment]
Any update on this? Is this planned?

**Comment by Unnamed2025** [comment]
It has already been implemented in the "Roo Code" and "CoolCline but they lack a restore feature.

**Comment by ViezeVingertjes** [comment]
This should fix it #1605 has been running for hours now without any rate limiting issue's. Left all the fancy stuff out though and went for fixing it functionally as it was barely useable for me without.

Using OpenRouter wasnt really a fix for me personally as OpenRouter is ALWAYS slow, so rather go with a fast API with proper retry policies.

**Comment by dcbartlett** [comment]
https://github.com/cline/cline/pull/1605 was merged and should be coming out soon.

**Comment by friedmud** [comment]
I see all of this work - but is there a way to configure the number of retries?  On Bedrock with a 429 it only retries twice... and way too quickly so they both fail.  I would like to tell it to retry more and pause longer.

**Comment by jingchang0623-crypto** [comment]
Auto-retry with exponential backoff for rate limits is a must-have. Here is our production setup:

**The 429 Nightmare** 

At 2:14 AM, my agent hit its 47th rate limit of the night. Not because it was doing anything wrong -- it was just TOO productive.

We run 5 agents 24/7 on OpenClaw, and rate limit handling was the first thing we had to solve. Our approach:

1. **Exponential backoff**: Start at 2s, double up to 60s max
2. **Token budget tracking**: Before each API call, check remaining tokens in the current window  
3. **Queue-based throttling**: If 3+ agents hit limits simultaneously, serialize requests
4. **Circuit breaker**: After 5 consecutive 429s, pause the agent for 5 minutes and notify

The key insight: rate limits are not errors, they are signals. Your agent should treat a 429 as the API saying "slow down, you are doing too much too fast" -- which is honestly a compliment.

More agent automation patterns: https://miaoquai.com/stories/
